Can't unpickle object overloaded by method

## Dotty 0.20.0 Source
```scala
object ObjOverload {
  def foo[A <: 42](a: A) = 42
  object foo { def apply(i: 43) = i }
}
```
## Scala 2.13.2 Source
```scala
object Test {
  def test = ObjOverload.foo(42)
}
```
## Crash Compiling Scala 2.13.2 Source
<details>
<summary>trace using -Ydebug-tasty</summary>

```scala
#[class ObjOverload]: Template: indexing members of object ObjOverload
#[class ObjOverload]: creating symbol <init> at Addr(60) with flags METHOD
#[class ObjOverload]: registered Symbol(constructor ObjOverload, #786041152) in object ObjOverload at Addr(60)
#[class ObjOverload]: creating symbol writeReplace at Addr(69) with flags PRIVATE | METHOD | SYNTHETIC
#[class ObjOverload]: registered Symbol(method writeReplace, #897074030) in object ObjOverload at Addr(69)
#[class ObjOverload]: creating symbol foo at Addr(92) with flags METHOD
#[class ObjOverload]: registered Symbol(method foo, #1885996206) in object ObjOverload at Addr(92)
#[class ObjOverload]: creating symbol foo at Addr(122) with flags FINAL | MODULE | STABLE | LAZY
#[class ObjOverload]: registered Symbol(object foo, #1859039536) in object ObjOverload at Addr(122)
#[class ObjOverload]: creating symbol foo at Addr(142) with flags FINAL | MODULE
error: java.lang.AssertionError: assertion failed: unpickling module class from TASTy before its module val.
	at scala.Predef$.assert(Predef.scala:282)
	at scala.tools.nsc.tasty.TreeUnpickler$TreeReader.createMemberSymbol(TreeUnpickler.scala:630)
	at scala.tools.nsc.tasty.TreeUnpickler$TreeReader.createSymbol(TreeUnpickler.scala:537)
	at scala.tools.nsc.tasty.TreeUnpickler$TreeReader.symbolAtCurrent(TreeUnpickler.scala:257)
	at scala.tools.nsc.tasty.TreeUnpickler$TreeReader.indexStats(TreeUnpickler.scala:774)
	at scala.tools.nsc.tasty.TreeUnpickler$TreeReader.readTemplate(TreeUnpickler.scala:968)
	at scala.tools.nsc.tasty.TreeUnpickler$TreeReader.readNewMember(TreeUnpickler.scala:917)
	at scala.tools.nsc.tasty.TreeUnpickler$TreeReader.readIndexedMember(TreeUnpickler.scala:840)
	at scala.tools.nsc.tasty.TreeUnpickler$Completer.$anonfun$complete$1(TreeUnpickler.scala:120)
	at scala.tools.nsc.tasty.bridge.ContextOps$Contexts$.withPhaseNoLater(ContextOps.scala:156)
	at scala.tools.nsc.tasty.TreeUnpickler$Completer.complete(TreeUnpickler.scala:119)
	...
```
</details>

--- a/src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala
+++ b/src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala
@@ -626,9 +626,8 @@ class TreeUnpickler[Tasty <: TastyUniverse](
               rootd
             case _ =>
               if (isModuleClass) {
-                val module = ctx.owner.rawInfo.decls.lookup(name.toTermName)
-                assert(module.isModule, "unpickling module class from TASTy before its module val.")
-                val moduleClass           = module.moduleClass
+                val moduleName            = name.toTermName
+                val moduleClass           = ctx.unlinkModule(moduleName).moduleClass
                 moduleClass.info          = completer
                 moduleClass.flags         = flags
                 moduleClass.privateWithin = privateWithin
@@ -668,8 +667,8 @@ class TreeUnpickler[Tasty <: TastyUniverse](
       sym
     }
 
-    private def allowsOverload(sym: Symbol) = ( // taken from Namer
-      sym.isSourceMethod && sym.owner.isClass && !sym.isTopLevel
+    private def allowsOverload(sym: Symbol) = ( // taken from Namer. TODO tasty: added module to allows overload
+      (sym.isSourceMethod || sym.isModule) && sym.owner.isClass && !sym.isTopLevel
     )
 
     /** Read modifier list into triplet of flags, annotations and a privateWithin

--- a/src/compiler/scala/tools/nsc/tasty/bridge/ContextOps.scala
+++ b/src/compiler/scala/tools/nsc/tasty/bridge/ContextOps.scala
@@ -2,6 +2,7 @@ package scala.tools.nsc.tasty.bridge
 
 import scala.annotation.tailrec
 
+import scala.collection.mutable
 import scala.reflect.io.AbstractFile
 import scala.tools.nsc.tasty.TastyUniverse
 
@@ -49,6 +50,13 @@ trait ContextOps extends TastyKernel { self: TastyUniverse =>
       def EmptyPackage: ModuleSymbol = loadingMirror.EmptyPackage
       def RootPackage: ModuleSymbol = loadingMirror.RootPackage
 
+      private[this] val modules: mutable.AnyRefMap[TermName, ModuleSymbol] = mutable.AnyRefMap.empty
+
+      def registerModule(moduleSym: ModuleSymbol): Unit = modules += moduleSym.name -> moduleSym
+      def unlinkModule(moduleName: TermName): ModuleSymbol =
+        modules.remove(moduleName).getOrElse(throw new AssertionError(
+          "unpickling module class from TASTy before its module val."))
+
       final lazy val loadingMirror: Mirror = initialContext.baseLoadingMirror
       final lazy val classRoot: Symbol = initialContext.baseClassRoot
 
@@ -68,7 +76,9 @@ trait ContextOps extends TastyKernel { self: TastyUniverse =>
             owner.newConstructor(noPosition, flags & ~Stable)
           }
           else if (flags.is(Module)) {
-            owner.newModule(name.toTermName, noPosition, flags)
+            val moduleSym = owner.newModule(name.toTermName, noPosition, flags)
+            registerModule(moduleSym)
+            moduleSym
           }
           else if (name.isTypeName) {
             owner.newTypeSymbol(name.toTypeName, noPosition, flags)

--- a/src/compiler/scala/tools/nsc/tasty/bridge/SymbolOps.scala
+++ b/src/compiler/scala/tools/nsc/tasty/bridge/SymbolOps.scala
@@ -32,7 +32,7 @@ trait SymbolOps extends TastyKernel { self: TastyUniverse =>
   }
 
   def selectSymFromSig(qualType: Type, name: Name, sig: Signature[Type])(implicit ctx: Context): Option[(Int, Symbol)] = {
-    ctx.log(s"""looking for overloaded method [$qualType]("$name") @@ ${sig.show}""")
+    ctx.log(s"""looking for overload member[$qualType]("$name") @@ ${sig.show}""")
     val MethodSignature(args, ret) = sig
     var seenTypeParams = false
     val member = qualType.member(name)
@@ -43,16 +43,19 @@ trait SymbolOps extends TastyKernel { self: TastyUniverse =>
       }
       (tyParamCounts.headOption.getOrElse(0), params)
     }
-    val alts = member.asTerm.alternatives
-    alts.find { sym =>
-      val method = sym.asMethod
-      val params = method.paramss.flatten
-      method.returnType.erasure =:= ret &&
+    def compareSym(sym: Symbol): Boolean = sym match {
+      case sym: MethodSymbol =>
+        val params = sym.paramss.flatten
+        sym.returnType.erasure =:= ret &&
         params.length === argsSyms.length &&
         ((name === nme.CONSTRUCTOR && tyParamCount === member.owner.typeParams.length)
-          || tyParamCount === method.typeParams.length) &&
+          || tyParamCount === sym.typeParams.length) &&
         params.zip(argsSyms).forall { case (param, tpe) => param.tpe.erasure =:= tpe }
-    }.map(tyParamCount -> _)
+      case _ =>
+        ctx.log(s"""member[$qualType]("$name") ${showSym(sym)} is not a method""")
+        false
+    }
+    member.asTerm.alternatives.find(compareSym).map(tyParamCount -> _)
   }
 
   def showSym(sym: Symbol): String = s"Symbol($sym, #${sym.hashCode})"

--- a/src/compiler/scala/tools/nsc/tasty/bridge/TastyKernel.scala
+++ b/src/compiler/scala/tools/nsc/tasty/bridge/TastyKernel.scala
@@ -131,6 +131,7 @@ trait TastyKernel { self: TastyUniverse =>
   def emptyScope: Scope = symbolTable.EmptyScope
 
   type Symbol = symbolTable.Symbol
+  type MethodSymbol = symbolTable.MethodSymbol
   type TermSymbol = symbolTable.TermSymbol
   type ModuleSymbol = symbolTable.ModuleSymbol
   type ClassSymbol = symbolTable.ClassSymbol

--- a/test/tasty/run/src-2/tastytest/TestDefAnnots.scala
+++ b/test/tasty/run/src-2/tastytest/TestDefAnnots.scala
@@ -1,0 +1,7 @@
+package tastytest
+
+object TestDefAnnots extends Suite("TestDefAnnots") {
+
+  test(assert(DefAnnots.withArgAnnot1("arg") === ("arg" : Any)))
+
+}

--- a/test/tasty/run/src-2/tastytest/TestHello.scala
+++ b/test/tasty/run/src-2/tastytest/TestHello.scala
@@ -16,7 +16,6 @@ object TestHello extends Suite("TestHello") {
   test(assert(HelloWorld.func1(33) === 33))
   test(assert((HelloWorld.lzy: "lazy") === "lazy"))
   test(assert(HelloWorld.acceptsOnlyMsg3(HelloWorld.msg3) === "Hello, World!Hello, World!"))
-  test(assert(HelloWorld.withArgAnnot("arg") === ("arg" : Any)))
   // test(assert(HelloWorld.`<init>` === 157)) // wait until https://github.com/lampepfl/dotty/issues/7799
 
 }

--- a/test/tasty/run/src-2/tastytest/TestObjOverload.scala
+++ b/test/tasty/run/src-2/tastytest/TestObjOverload.scala
@@ -1,5 +1,6 @@
 package tastytest
 
 object TestObjOverload extends Suite("TestObjOverload") {
+  test(assert(ObjOverload.foo(43) === 43))
   test(assert(ObjOverload.foo(42) === 42))
 }

--- a/test/tasty/run/src-2/tastytest/TestObjOverload.scala
+++ b/test/tasty/run/src-2/tastytest/TestObjOverload.scala
@@ -1,0 +1,5 @@
+package tastytest
+
+object TestObjOverload extends Suite("TestObjOverload") {
+  test(assert(ObjOverload.foo(42) === 42))
+}

--- a/test/tasty/run/src-3/tastytest/DefAnnots.scala
+++ b/test/tasty/run/src-3/tastytest/DefAnnots.scala
@@ -1,0 +1,11 @@
+package tastytest
+
+import scala.annotation.StaticAnnotation
+
+object DefAnnots {
+
+  class argAnnot(arg: Any) extends StaticAnnotation
+
+  def withArgAnnot1(arg: Any @argAnnot(??? : Int)): Any = arg
+
+}

--- a/test/tasty/run/src-3/tastytest/HelloWorld.scala
+++ b/test/tasty/run/src-3/tastytest/HelloWorld.scala
@@ -4,8 +4,6 @@ import scala.annotation.StaticAnnotation
 
 object HelloWorld {
 
-  class argAnnot(arg: Any) extends StaticAnnotation
-
   val msg1: String = "Hello, World!"
   def msg2: String = "Hello, World!"
   final val msg3 = "Hello, World!"
@@ -21,6 +19,6 @@ object HelloWorld {
   def func1[A]: A => A = x => x
   def acceptsOnlyMsg3(m: msg3.type): String = m + m
   final lazy val lzy = "lazy"
-  def withArgAnnot(arg: Any @argAnnot(??? : Int)): Any = arg
   // def `<init>`: Int = 157 // broken in https://github.com/lampepfl/dotty/issues/7799
+
 }

--- a/test/tasty/run/src-3/tastytest/ObjOverload.scala
+++ b/test/tasty/run/src-3/tastytest/ObjOverload.scala
@@ -1,7 +1,5 @@
 package tastytest
 
-import scala.annotation.StaticAnnotation
-
 object ObjOverload {
 
   def foo[A <: 42](a: A) = 42

--- a/test/tasty/run/src-3/tastytest/ObjOverload.scala
+++ b/test/tasty/run/src-3/tastytest/ObjOverload.scala
@@ -1,0 +1,10 @@
+package tastytest
+
+import scala.annotation.StaticAnnotation
+
+object ObjOverload {
+
+  def foo[A <: 42](a: A) = 42
+  object foo { def apply(i: 43) = i }
+
+}

